registry/eureka/watcher.go

```diff
@@ -0,0 +1,296 @@
+			log.Info("eureka watcher(%v) is stopping ...", w.Name)
+			return
+		}
+	}
+}
+
+func (w *watcher) Stop() {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	for serviceName := range w.WatchingServices {
+		if err := w.unsubscribe(serviceName); err != nil {
+			log.Errorf("Failed to unsubscribe service : %v", serviceName)
+			continue
+		}
+		w.cache.DeleteServiceEntryWrapper(makeHost(serviceName))
+	}
+	w.UpdateService()
+}
+
+func (w *watcher) IsHealthy() bool {
+	return w.Status == provider.Healthy
+}
+
+func (w *watcher) GetRegistryType() string {
+	return w.RegistryType.String()
+}
+
+// doFullRefresh todo(lql): it's better to support deltaRefresh
+func (w *watcher) doFullRefresh() {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	applications, err := w.eurekaClient.GetApplications()
+	if err != nil {
+		log.Errorf("Failed to full fetch eureka services, error : %v", err)
+		return
+	}
+
+	fetchedServices := applications.Apps
+	for serviceName := range w.WatchingServices {
+		if _, ok := fetchedServices[serviceName]; !ok {
+			if err = w.unsubscribe(serviceName); err != nil {
+				log.Errorf("Failed to unsubscribe service %v, error : %v", serviceName, err)
+				continue
+			}
+		}
+	}
+
+	for serviceName := range fetchedServices {
+		if _, ok := w.WatchingServices[serviceName]; !ok {
+			if err = w.subscribe(fetchedServices[serviceName]); err != nil {
+				log.Errorf("Failed to subscribe service %v, error : %v", serviceName, err)
+				continue
+			}
+		}
+	}
+}
+
+func (w *watcher) subscribe(service *fargo.Application) error {
+	if service == nil {
+		return fmt.Errorf("service is nil")
+	}
+	callback := func(service *fargo.Application) error {
+		defer w.UpdateService()
+
+		if len(service.Instances) != 0 {
+			se := generateServiceEntry(service)
+			w.cache.UpdateServiceEntryWrapper(makeHost(service.Name), &memory.ServiceEntryWrapper{
+				ServiceName:  service.Name,
+				ServiceEntry: se,
+				Suffix:       suffix,
+				RegistryType: w.Type,
+			})
+			return nil
+		}
+
+		if w.updateCacheWhenEmpty {
+			w.cache.DeleteServiceEntryWrapper(makeHost(service.Name))
+		}
+
+		return nil
+	}
+
+	callback(service)
+	w.WatchingServices[service.Name] = NewPlan(w.eurekaClient, service.Name, callback)
+	return nil
+}
+
+func (w *watcher) unsubscribe(serviceName string) error {
+	w.WatchingServices[serviceName].Stop()
+	delete(w.WatchingServices, serviceName)
+	w.UpdateService()
+
+	return nil
+}
+
+func makeHost(serviceName string) string {
+	return serviceName + common.DotSeparator + suffix
+}
+
+func convertMap(m map[string]interface{}) map[string]string {
+	result := make(map[string]string)
+	for k, v := range m {
+		if value, ok := v.(string); ok {
+			result[k] = value
+		}
+	}
+
+	return result
+}
+
+func generateServiceEntry(app *fargo.Application) *v1alpha3.ServiceEntry {
+	portList := make([]*v1alpha3.Port, 0)
+	endpoints := make([]*v1alpha3.WorkloadEntry, 0)
+
+	for _, instance := range app.Instances {
+		protocol := common.HTTP
```

Comment:
We should support other protocols, such as gRPC, which you can parse from the service's metadata if present.

Reply:
maybe we can get protocol type from `instance.Metadata`

```
if val, _ := instance.Metadata.GetString("protocol"); val != "" {
			protocol = common.ParseProtocol(val)
		}
```

Reply:
That's ok

Reply:
You should check the protocol is valid, if not, should use the http as default.